File: node/bench.js
import pkg from 'kafkajs';
const { Kafka, CompressionTypes, CompressionCodecs, logLevel } = pkg;
import SnappyCodec from 'kafkajs-snappy';
// import LZ4Codec from 'kafkajs-lz4';
import avro from 'avsc';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

// Register compression codecs
CompressionCodecs[CompressionTypes.Snappy] = SnappyCodec;
// CompressionCodecs[CompressionTypes.LZ4] = new LZ4Codec().codec;

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// ============================================================================
// Configuration
// ============================================================================

let TEXT_MESSAGES = [
  'Hello world',
  'Quick brown fox jumps over the lazy dog',
  'Kafka benchmarking message',
  'Avro vs JSON throughput test',
  'Message compression experiment',
];

const argv = yargs(hideBin(process.argv))
  .options({
    mode: { type: 'string', choices: ['json', 'avro'], demandOption: true },
    messages: { type: 'number', default: 100000 },
    batch: { type: 'number', default: 100 },
    compression: { type: 'string', choices: ['none', 'gzip', 'snappy', 'lz4'], default: 'gzip' },
    brokers: { type: 'string', default: 'localhost:9094' },
    topic: { type: 'string', default: '' },
    large: { type: 'boolean', default: false },
  })
  .parseSync();

// ============================================================================
// Helper Functions
// ============================================================================

function getCompressionType(name) {
  const map = {
    'gzip': CompressionTypes.GZIP,
    'snappy': CompressionTypes.Snappy,
    'lz4': CompressionTypes.LZ4,
    'none': CompressionTypes.None,
  };
  return map[name] || CompressionTypes.None;
}

function buildRecord(i) {
  return {
    id: String(i),
    ts_unix_ms: Date.now(),
    category: 'bench',
    value: Math.random(),
    message: TEXT_MESSAGES[i % TEXT_MESSAGES.length],
  };
}

function serializeRecord(record, mode, avroType) {
  if (mode === 'json') {
    return Buffer.from(JSON.stringify(record));
  } else {
    return avroType.toBuffer(record);
  }
}

function deserializeRecord(buffer, mode, avroType) {
  if (mode === 'json') {
    return JSON.parse(buffer.toString('utf8'));
  } else {
    return avroType.fromBuffer(buffer);
  }
}

// ============================================================================
// Main Benchmark
// ============================================================================

async function runBenchmark() {
  const { mode, messages: totalMessages, batch: batchSize, compression, brokers, large } = argv;

  const topic = argv.topic || (mode === 'json' ? 'bench-json' : 'bench-avro');
  const compressionType = getCompressionType(compression);

  if (large) {
    TEXT_MESSAGES = TEXT_MESSAGES.map(message => message.repeat(1000));
  }

  // Load Avro schema if needed
  let avroType = null;
  if (mode === 'avro') {
    const schemaPath = path.join(__dirname, '../schemas/event.avsc');
    const schemaJson = JSON.parse(fs.readFileSync(schemaPath, 'utf8'));
    avroType = avro.Type.forSchema(schemaJson);
  }

  // Setup Kafka client
  const kafka = new Kafka({
    brokers: brokers.split(','),
    logLevel: logLevel.NOTHING,
  });

  // ============================================================================
  // Step 1: Start Consumer (subscribe from LATEST, not beginning)
  // ============================================================================

  const consumer = kafka.consumer({
    groupId: `bench-node-${mode}-${Date.now()}`,
  });
  await consumer.connect();
  
  // Subscribe from latest - we only want NEW messages from this run
  await consumer.subscribe({ topic, fromBeginning: false });

  let consumedCount = 0;
  let totalMessageLength = 0;
  let consumeStart = null;
  let consumeEnd = null;
  let resolveConsumption;
  const consumptionComplete = new Promise(resolve => {
    resolveConsumption = resolve;
  });

  // Start consuming in background
  consumer.run({
    eachMessage: async ({ message }) => {
      if (!message.value) return;
      
      // Mark consumption start time on first message
      if (consumeStart === null) {
        consumeStart = Date.now();
      }
      
      try {
        const record = deserializeRecord(message.value, mode, avroType);
        totalMessageLength += record.message.length;
        consumedCount++;

        // Check if we've consumed all messages
        if (consumedCount >= totalMessages) {
          consumeEnd = Date.now();
          resolveConsumption();
        }
      } catch (err) {
        // Skip malformed messages
      }
    },
  });

  // Give consumer time to subscribe and be ready
  await new Promise(resolve => setTimeout(resolve, 2000));

  // ============================================================================
  // Step 2: Produce Messages
  // ============================================================================

  const producer = kafka.producer();
  await producer.connect();

  const produceStart = Date.now();
  let producedCount = 0;
  let messageBatch = [];

  for (let i = 0; i < totalMessages; i++) {
    const record = buildRecord(i);
    const value = serializeRecord(record, mode, avroType);
    messageBatch.push({ value });

    // Send batch when full
    if (messageBatch.length >= batchSize) {
      await producer.send({ topic, compression: compressionType, messages: messageBatch });
      producedCount += messageBatch.length;
      messageBatch = [];
    }
  }

  // Send remaining messages
  if (messageBatch.length > 0) {
    await producer.send({ topic, compression: compressionType, messages: messageBatch });
    producedCount += messageBatch.length;
  }

  const produceEnd = Date.now();
  await producer.disconnect();

  // ============================================================================
  // Step 3: Wait for Consumption to Complete
  // ============================================================================

  await consumptionComplete;
  await consumer.disconnect();

  // ============================================================================
  // Step 4: Report Results
  // ============================================================================

  const produceDuration = (produceEnd - produceStart) / 1000;
  const produceRate = producedCount / produceDuration;
  const consumeDuration = consumeEnd && consumeStart ? (consumeEnd - consumeStart) / 1000 : 0;

  const result = {
    lang: 'node',
    mode,
    topic,
    produced: producedCount,
    durationSec: produceDuration,
    rate: produceRate,
    consumed: consumedCount,
    stringLenSum: totalMessageLength,
    consumeDurSec: consumeDuration,
  };

  console.log(JSON.stringify(result, null, 2));
  
  // Ensure process exits
  process.exit(0);
}

// ============================================================================
// Entry Point
// ============================================================================

runBenchmark().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});

File: python/bench.py
#!/usr/bin/env python3
"""
Kafka benchmark: JSON vs Avro serialization
Simple, linear workflow for easy verification
"""

import time
import json
import io
import argparse
from pathlib import Path
from confluent_kafka import Producer, Consumer
from fastavro import schemaless_writer, schemaless_reader
import orjson

# ============================================================================
# Configuration
# ============================================================================

TEXT_MESSAGES = [
    "Hello world",
    "Quick brown fox jumps over the lazy dog",
    "Kafka benchmarking message",
    "Avro vs JSON throughput test",
    "Message compression experiment",
]


# ============================================================================
# Helper Functions
# ============================================================================

def build_record(i: int) -> dict:
    """Build a single benchmark record"""
    return {
        "id": str(i),
        "ts_unix_ms": int(time.time() * 1000),
        "category": "bench",
        "value": float(i % 1000) / 1000.0,
        "message": TEXT_MESSAGES[i % len(TEXT_MESSAGES)],
    }


def load_avro_schema() -> dict:
    """Load Avro schema from file"""
    schema_path = Path(__file__).resolve().parents[1] / "schemas" / "event.avsc"
    with open(schema_path) as f:
        return json.load(f)


_bytesio = io.BytesIO()

def serialize_record(record: dict, mode: str, schema: dict = None) -> bytes:
    """Serialize record to JSON or Avro"""
    if mode == "json":
        return orjson.dumps(record)
    else:
        _bytesio.seek(0)
        _bytesio.truncate(0)
        schemaless_writer(_bytesio, schema, record)
        return _bytesio.getvalue()


def deserialize_record(data: bytes, mode: str, schema: dict = None) -> dict:
    """Deserialize record from JSON or Avro"""
    if mode == "json":
        return json.loads(data.decode("utf-8"))
    else:
        return schemaless_reader(io.BytesIO(data), schema)


def sum_message_lengths(record: dict) -> int:
    """Calculate total length of message field"""
    return len(record.get("message", ""))


# ============================================================================
# Main Benchmark
# ============================================================================

def main():
    # Parse arguments
    parser = argparse.ArgumentParser(description="Kafka JSON vs Avro benchmark")
    parser.add_argument("--mode", choices=["json", "avro"], required=True)
    parser.add_argument("--messages", type=int, default=100000)
    parser.add_argument("--batch", type=int, default=100)
    parser.add_argument("--compression", default="gzip")
    parser.add_argument("--brokers", default="localhost:9094")
    parser.add_argument("--topic", default="")
    parser.add_argument("--large", action="store_true")
    args = parser.parse_args()

    topic = args.topic or (f"bench-{args.mode}")
    schema = load_avro_schema() if args.mode == "avro" else None

    global TEXT_MESSAGES
    if args.large:
        TEXT_MESSAGES = [text * 1000 for text in TEXT_MESSAGES]

    # ============================================================================
    # Step 1: Start Consumer (before producing)
    # ============================================================================

    consumer_config = {
        "bootstrap.servers": args.brokers,
        "group.id": f"bench-python-{args.mode}-{int(time.time() * 1000)}",
        "auto.offset.reset": "earliest",
        "enable.auto.commit": True,
    }
    consumer = Consumer(consumer_config)
    consumer.subscribe([topic])

    consumed_count = 0
    total_message_length = 0
    consume_start = None
    consume_end = None

    # ============================================================================
    # Step 2: Produce Messages (with concurrent consumption)
    # ============================================================================

    producer_config = {
        "bootstrap.servers": args.brokers,
        "compression.type": args.compression,
        "batch.num.messages": args.batch,
    }
    producer = Producer(producer_config)

    produce_start = time.time()
    produced_count = 0

    for i in range(args.messages):
        record = build_record(i)
        value = serialize_record(record, args.mode, schema)
        producer.produce(topic=topic, value=value)
        produced_count += 1
        
        # Poll producer to handle delivery callbacks
        if produced_count % args.batch == 0:
            producer.poll(0)

    # Flush all remaining messages
    producer.flush()
    produce_end = time.time()

    # ============================================================================
    # Step 3: Consume Until All Messages Received
    # ============================================================================

    # Poll for messages until we've consumed everything we produced
    while consumed_count < produced_count:
        msg = consumer.poll(timeout=1.0)
        
        if msg is None:
            continue
        if msg.error():
            continue
        
        value = msg.value()
        if not value:
            continue
        
        # Mark consumption start time on first message
        if consume_start is None:
            consume_start = time.time()
        
        try:
            record = deserialize_record(value, args.mode, schema)
            total_message_length += sum_message_lengths(record)
            consumed_count += 1
        except Exception:
            # Skip malformed messages
            continue

    consume_end = time.time()
    consumer.close()

    # ============================================================================
    # Step 4: Report Results
    # ============================================================================

    produce_duration = produce_end - produce_start
    produce_rate = produced_count / produce_duration if produce_duration > 0 else 0
    consume_duration = (consume_end - consume_start) if consume_start and consume_end else 0

    result = {
        "lang": "python",
        "mode": args.mode,
        "topic": topic,
        "produced": produced_count,
        "durationSec": produce_duration,
        "rate": produce_rate,
        "consumed": consumed_count,
        "stringLenSum": total_message_length,
        "consumeDurSec": consume_duration,
    }

    print(json.dumps(result, indent=2))


if __name__ == "__main__":
    main()